src/app/AttributePathExpandIterator.cpp

```diff
@@ -162,14 +156,28 @@ std::optional<AttributeId> AttributePathExpandIterator::NextAttributeId()
             //
             // For wildcard expansion, we validate that this is a valid attribute for the given
             // cluster on the given endpoint. If not a wildcard expansion, return it as-is.
-            if (mPosition.mAttributePath->mValue.IsWildcardPath())
+            DataModel::AttributeFinder finder(mDataModelProvider);
+
+            const ConcreteAttributePath attributePath(mPosition.mOutputPath.mEndpointId, mPosition.mOutputPath.mClusterId,
+                                                      mPosition.mAttributePath->mValue.mAttributeId);
+            std::optional<DataModel::AttributeEntry> entry = finder.Find(attributePath);
+
+            if (entry.has_value() || mPosition.mAttributePath->mValue.IsWildcardPath())
             {
-                if (!IsValidAttributeId(mPosition.mAttributePath->mValue.mAttributeId))
-                {
-                    return std::nullopt;
-                }
+                return entry;
             }
-            return mPosition.mAttributePath->mValue.mAttributeId;
+
+            // We get here if:
+            //   - entry is NOT valid (this is not a valid attribute)
+            //   - path is NOT a wildcard (i.e. we were asked to explicitly return it)
+            // as a result, we have no way to generate a "REAL" attribute metadata, so return one
+            // that is not accessible at all
```

Comment:
This suggests that our API is broken.

All this iterator is supposed to return is paths.  Then something else should be fetching the metadata for those paths as needed, and doing the right things with it.

In particular, per spec ... well, I have no idea what should be happening here, because I don't think the spec as written is implementable.  So we better get that sorted out before we start introducing new behaviors here for clients to deal with.

Reply:
Our tests explicitly test "if I give you a path without wildcards, you shall return this path for me". I believe this would be the equivalent of the spec requesting a read of a path that is invalid.

This is existing behavior rather than previous, the only danger seems to be if we use the "access data" where we previously did not. However since access is "denied" now that seems to be reasonable.

Reply:
Yes, but my point is that returning "no read or write privileges" for such a path is not correct.

Reply:
ok. For now std::nullopt should work, will just validate the rest of the paths.